Add eInk plugin

jano42 opened the pull request:
Allow to control eInk screen through HTTP API

Nearly same plugin as https://github.com/letscontrolit/ESPEasy/pull/2693

![image](https://user-images.githubusercontent.com/3135683/80229773-32bde480-8651-11ea-9824-cf37c606534f.png)


## INTRO

This plugin allow to control a eInk screen (ILI3897) through HTTP API

## Environment
Tested with Lolin d32 pro and Wemos ePaper 2.13 shield
Tested with last ESPEasy Release mega-20200410 on ESP32

ePaper Shield : https://www.wemos.cc/en/latest/d1_mini_shiled/epd_2_13.html
Price : ~ 10€/$ (https://fr.aliexpress.com/item/32981318996.html)


## Dependencies
Plugin lib_deps = Adafruit GFX, LOLIN_EPD

## API Documentation

This plugin is controlled by HTTP API, ie : http://[espeasy_ip]/control?cmd=epd,tx,HelloWorld

| command | details | description |
|-----|-----|-----|
| epd | `EPD,[epd_subcommand],....` | Draw line, rect, circle, triangle and text |
|epdcmd | `EPDCMD,[epdcmd_subcommand]` | Control the screen (on, off, clear,..) |

EPD Subcommands:

| EPD Subcommands | details | description |
|-----|-----|-----|
| txt | txt,[text] | Write simple text (use last position, color and size) |
| txp | txp,[X],[Y] | Set text position (move the cursor) |
| txc | txc,[foreColor],[backgroundColor] | Set text color (background is transparent if not provided |
| txz | txz,[X],[Y],[text] | Write text at position (move the cursor + print) |
| txs | txs,[SIZE] | Set text size |
| txtfull | txtfull,[row],[col],[size=1],[foreColor=white],[backColor=black],[text] | Write text with all options |
| l | l,[x1],[y1],[2],[y2],[color] | Draw a simple line |
| lh | lh,[y],[width],[color] | Draw an horizontal line (width = Line width in pixels (positive = right of first point, negative = point of first corner). |
| lv | lv,[x],[height],[color] | Draw a vertical line (height= Line height in pixels (positive = below first point, negative = above first point).|
| r | r,[x],[y],[width],[height],[color] | Draw a rectangle |
| rf | rf,[x],[y],[width],[height],[bordercolor],[innercolor] | Draw a filled rectangle |
| c | c,[x],[y],[radius],[color] | Draw a circle |
| cf | cf,[x],[y],[radius],[bordercolor],[innercolor] | Draw a filled circle |
| t | t,[x1],[y1],[x2],[y2],[x3],[y3],[color]| Draw a triangle |
| tf | tf,[x1],[y1],[x2],[y2],[x3],[y3],[bordercolor],[innercolor] | Draw a filled triangle |
| rr | rr,[x],[y],[width],[height],[corner_radius],[color] | Draw a round rectangle |
| rrf | rrf,[x],[y],[width],[height],[corner_radius],[bordercolor],[innercolor] | Draw a filled round rectangle |
| px | px,[x],[y],[color] | Print a single pixel |
  
EPDCMD Subcommands:

| EPD Subcommands | details | description |
|-----|-----|-----|
| clear | clear,[color] | Clear display |
| deepsleep | deepsleep | Make screen go to sleep |
| inv | inv,[value] | Invert the dispaly (value:0 normal display, 1 inverted display) |
| rot | rot,[value] | Rotate display (value from 0 to 3 inclusive) |


Examples:

	Write Text :
		http://[espeasy_ip]/control?cmd=epd,txtfull,0,0,HelloWorld
	
	Write Text another place:
		http://[espeasy_ip]/control?cmd=epd,txtfull,100,40,HelloWorld

	Write bigger Text :
		http://[espeasy_ip]/control?cmd=epd,txtfull,0,0,3,HelloWorld

	Write RED Text :
		http://[espeasy_ip]/control?cmd=epd,txtfull,0,0,3,HelloWorld

	Write RED Text (size is 1):
		http://[espeasy_ip]/control?cmd=epd,txtfull,0,0,1,RED,HelloWorld

	Write RED Text on YELLOW background (size is 1):
		http://[espeasy_ip]/control?cmd=epd,txtfull,0,0,1,RED,YELLOW,HelloWorld
		
	Clear whole display
		http://[espeasy_ip]/control?cmd=epdcmd,clear

	Deepsleep screen
		http://[espeasy_ip]/control?cmd=epdcmd,deepsleep

## Stats

RAM:   [==        ]  17.7% (used 58024 bytes from 327680 bytes)
Flash: [==========]  95.1% (used 1806626 bytes from 1900544 bytes)

jano42 committed:
Add eInk plugin

Allow to control eInk screen through HTTP API

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+    case PLUGIN_DEVICE_ADD:
+      {
+        Device[++deviceCount].Number = PLUGIN_ID_096;
+        Device[deviceCount].Type = DEVICE_TYPE_I2C;
```
It is not an I2C device, but SPI.
Currently we don't have a SPI type definition, but I guess we could need one.
Maybe you can add that type also?

jano42 replied:
ok done

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+        if(!success)
+        {
+          if (loglevelActiveFor(LOG_LEVEL_INFO)) {
+            addLog(LOG_LEVEL_INFO, F("Fail to parse command correctly; please check API documentation"));
```
This will give lots and lots of log entries, if commands are given for other tasks/plugins.

`success = false` is something that will happen in almost all calls to `PLUGIN_WRITE`  (except  for the task the command was meant to be sent)

jano42 replied:
ok you're right. I've removed those log lines.

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+        log.reserve(110);                           // Prevent re-allocation
+        log = F("P096-eInk : WRITE = ");
+        log += tmpString;
+        SendStatus(event->Source, log);             // Reply (echo) to sender. This will print message on browser.  
```
This is also printed when `!success`

jano42 replied:
I just move this code in #ifdef DEBUG_LOG (to allow easy debugging in case of evolutions)

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+  String stringOne  = string;
+  String stringTwo = "";
+  stringTwo += (char)247;
+  stringOne.replace("°", stringTwo);
```
Do not use extended ASCII or Unicode in source code (not even in comments)

jano42 replied:
Ok I've done this, to allow displaying the "degree" symbol.
The degree symbol is not in standard font and cannot be displayed (print("°") display strange cars)
As my personal main usage is for domotic software, degrees is a very important car.
Is there any other way to do it ?

TD-er replied:
I do have similar replacements in the file StringConverter.ino.
See:

```c++
  {
    // Degree
    const char degree[3]   = { 0xc2, 0xb0, 0 };       // Unicode degree symbol
    const char degreeC[4]  = { 0xe2, 0x84, 0x83, 0 }; // Unicode degreeC symbol
    const char degree_C[4] = { 0xc2, 0xb0, 'C', 0 };  // Unicode degree symbol + captial C
    repl(F("{D}"),   degree,   s, useURLencode);
    repl(F("&deg;"), degree,   s, useURLencode);
    repl(degreeC,    degree_C, s, useURLencode);
  }
```

jano42 replied:
OK thanks.
Done.

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+    return EPD_DARK;
+  if (s.equalsIgnoreCase(F("LIGHT")))
+    return EPD_LIGHT;
+  return EPD_WHITE; //fallback value
```
These calls do compare a lot with `equalsIgnoreCase`.
Maybe make a local copy of the string, convert it to lower case and compare with `.equals` on that lower case string.
Compares with ignore case are quite resource intensive.

jano42 replied:
ok

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -0,0 +1,603 @@
+  return EPD_WHITE; //fallback value
+}
+
+String Plugin_096_FixText(String & s)
```
You're making a copy of the string, so the argument can be a `const String& s`  (also for other functions that do not intend to change the parameter)

jano42 replied:
Quite good remark ;-)

TD-er commented:
Do you also have some timing stats of this plugin?
E-ink is quite slow, so I wonder how long it takes to update the screen.

jano42 committed:
Add SPI device type

jano42 committed:
Optimize code:
 - remove unwanted output
- move http answer only in debug mode
- fix functions args to const
- optimize parsing color data

TD-er commented on src/_P096_eInk.ino:
```diff
@@ -279,7 +279,9 @@ boolean Plugin_096(byte function, struct EventStruct *event, String& string)
 
     case PLUGIN_WRITE:
       {
+#ifdef DEBUG_LOG
```
We already have `BUILD_NO_DEBUG`
So please change into `#ifndef BUILD_NO_DEBUG`

jano42 replied:
OK I made a mistake (I've searched in espeasy code, and found somewhere DEBUG_LOG, so I've decided to use it)
Fixed ;-)

jano42 commented:
I've made some tests, and updating the screen (complete erase and display)
is about 1 sec.
The e-Ink screen is not usually used for performance, but more for low
power consumption (and deepsleep mode, which make screen keep displayed
without any power)

Le ven. 24 avr. 2020 à 20:34, TD-er <notifications@github.com> a écrit :

> Do you also have some timing stats of this plugin?
> E-ink is quite slow, so I wonder how long it takes to update the screen.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letscontrolit/ESPEasy/pull/3033#issuecomment-619177846>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAX5RQ57TN4CPUFAOEGBJIDROHLUJANCNFSM4MQGFIMA>
> .
>

jano42 committed:
Remove useless comments
Use BUILD_NO_DEBUG define
Add display timings
Fix input text to handle degress and monetary symbols (specific for adafruit gfx lib with eink screen)

TD-er commented:
I know how e-ink works, that's why I asked.

Do you think the transfer of data to the e-ink display can be split into separate parts?
Is it possible a transfer may take longer than 1 second? (watchdog)
What is the effect of running background tasks during update?

tonhuisman commented:
> r 	r,[x],[y],[width],[height],[color] 	Draw a rectangle
> rf 	r,[x],[y],[width],[height],[bordercolor],[innercolor] 	Draw a filled rectangle

I notice quite a few typos in the commands documentation, above is just an example (sub-command example for 'rf' shows 'r' instead). It would be nice to have those corrected to avoid disappointment or bug reports of users of the plugin (possibly me).

TD-er committed:
Merge branch 'mega' into mega

jano42 committed:
Fix documentation typo

jano42 committed:
Merge branch 'mega' of https://github.com/jano42/ESPEasy into mega

jano42 commented:
@TD-er : I'll see the IL3897 documentation in details and the actual implementation. Actually the display method seems locking for about 1 second (so background task may be blocked as well). I'll see and tell you about that.

@tonhuisman : 👍 I've fixed it , in code and also in this PR description

TD-er commented:
Are these typos also present in the TFT plugin you made?
I already corrected one in the comments of that plugin when I copied it to the docs.
So if there are typos present in that plugin as well, please also change them in the docs.

TD-er commented:
Travis reports this error:
```
/home/travis/build/letscontrolit/ESPEasy/src/_P096_eInk.ino:100:23: fatal error: LOLIN_EPD.h: No such file or directory
```

jano42 commented:
> Travis reports this error:
> 
> ```
> /home/travis/build/letscontrolit/ESPEasy/src/_P096_eInk.ino:100:23: fatal error: LOLIN_EPD.h: No such file or directory
> ```

The file is in a dependency library.
I'm not comfortable with pio deps and how to fix it. I've just added LOLIN_EPD to my env.
Is there something else to do ? put it in every env ?

TD-er commented:
If you need it for all, then place it in the libdeps declaration where also ESPEasySerial is defined. (that one is also defined for ESP82xx and ESP32)

jano42 committed:
Fix documentation typo

jano42 commented:
> Are these typos also present in the TFT plugin you made?
> I already corrected one in the comments of that plugin when I copied it to the docs.
> So if there are typos present in that plugin as well, please also change them in the docs.

I've checked and the same errors are presents (logical since I've copy pasted plugin).
I've pushed update.

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er commented:
The last failed build was failing because it tries to include a file named `LOLIN_EPD.h` which is not included in this PR.

TD-er committed:
[e-ink] Include LOLIN_EPD in the lib_deps to make it compile

TD-er committed:
Merge branch 'mega' into mega

JohnsonLv commented:
I use Rules  but can not display BME280 context in the same screen 
on System#Boot do
  epaper,txtfull,1,1,4,hello!:)
   timerSet,1,5
endon

On Rules#Timer=1  Do
 epaper,txtfull,1,1,2,"[BME280#Temperature] {D}C" 
  epaper,txtfull,1,2,4," [BME280#Humidity]%" 
  timerSet,1,5
Endon
how to display temperature and humidity at the same time

tonhuisman commented:
You've found a _very_ old, and already closed, pull request to hide your question 😨, this could have better been asked in [the forum](https://www.letscontrolit.com/forum)

> how to display temperature and humidity at the same time

The documentation for this plugin is a bit hard to find, as that's partly still to be written 😱 
An exception to that is that it's using the AdafruitGFX_Helper module, using a shared set of commands across many of the display plugins in ESPEasy, as documented f.e. in [P095](https://espeasy.readthedocs.io/en/latest/Plugin/P095.html#commands-available) (TFT), but using the `epaper` and `epapercmd` triggers instead of the triggers listed there, and of course only the colors black and white can be used.
Updating an epaper display is a somewhat special process, and in fact updates the entire display for each separate command.
To avoid that, you can enter the desired layout in the Lines, available in the plugin configuration,
or use the `epaper,seq_start` command, fill the display using the available text and drawing commands, and finally send the command `epaper,seq_end`, to update the entire display 😃

JohnsonLv commented:
Hello Tonhuisman,
Thank very much, because I want to set different font in different line !
I tried below rules ,it is still not OK
on System#Boot do
epaper,txtfull,1,1,4,hello!:)
timerSet,1,5
endon

On Rules#Timer=1 Do
epaper,seq_start
epaper,txtfull,1,1,2,"[BME280#Temperature] {D}C"
epaper,txtfull,1,2,4," [BME280#Humidity]%"
epaper,seq_end
timerSet,1,5
Endon

tonhuisman commented:
You haven't clearly stated what _is_ and what _isn't_ working as expected, can you be more elaborate?

Btw, updating your eInk display very 5 seconds is not a very good plan. For displaying rapidly changing content, an OLed, LCD or TFT display is much more suitable, as changing the content very often will, over time, damage the eInk display.

As said, with such type of content, it is better suited to put that data in the Lines you find at the Device Configuration of the task. Suggested interval should be at least 300 seconds (every 5 minutes) or possibly 600 (every 10 minutes).

JohnsonLv commented:
![image](https://github.com/letscontrolit/ESPEasy/assets/13288567/2dc352d0-d392-4133-8f03-55d257db7b12)
Actually ,I'd like to know how to achieve the same effect as the image above？
Thank you in advanced!
